### PR TITLE
Ldap hide server manager password

### LINKED ISSUE #4241: LDAP server manager password visible and stored in plain text
Setting up a LDAP server the server manager password is stored in plain text in the database and is visible as plain text if you open this entry!

### PR DESCRIPTION
**Pull request by markusweigelt**
Fixes #4241.

-  Adding the encryption and decryption functionalities
- Migration of the existing server manager passwords under System -> Migration -> Migrate LDAP manager passwords if secret param is set in kitodo_config.properties (can run several times cause already encrypted passwords will be ignored)

### CODE REVIEW

**Comment by henning-gerhardt** [comment]
I need some time to review and test this pull request as I did not have a development system with Java 11.

File: Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package org.kitodo.production.security;
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java:1]: Missing licence

File: Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java [hunk 2]
```diff
@@ -333,8 +325,8 @@ public String createNewWorkflow() {
     }
 
     /**
-     * When the navigation to the migration form is coming from a workflow
-     * creation the URL contains a WorkflowId.
+     * When the navigation to the migration form is coming from a workflow creation
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java:328]: there is again some formatting changes

File: Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java [hunk 3]
```diff
@@ -13,28 +13,19 @@
...
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
+import java.util.*;
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java:16]: please avoid * imports

File: Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            return null;
+        }
+
+        KeySpec spec = new PBEKeySpec(secret.toCharArray(), salt, 65536, 256); // AES-256
+        SecretKeyFactory f = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
+        byte[] key = f.generateSecret(spec).getEncoded();
+        return new SecretKeySpec(key, "AES");
+    }
+
+    public static String encrypt(String input, String secret)
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java:45]: Javadoc missing (for all public methods here)

File: Kitodo/src/test/java/org/kitodo/production/security/AESUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package org.kitodo.production.security;
```
Comment (Kathrin-Huber) [inline Kitodo/src/test/java/org/kitodo/production/security/AESUtilTest.java:1]: Licence header missing

**Review by Kathrin-Huber — CHANGES_REQUESTED** [review]

File: Kitodo/src/main/java/org/kitodo/config/enums/ParameterCore.java [hunk 6]
```diff
@@ -610,9 +610,19 @@
...
+
+    /*
+     * Security properties
+     */
+
+    /*
+     * Secret is used to encrypt or decrypt sensitive data e.g. LDAP manager passwords which are stored in the database in encrypted form.
+     * Once the secret is set, it should not be changed since encrypted data can no longer be decrypted.
+     */
+    SECURITY_SECRET(new Parameter<>("security.secret", ""));
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/config/enums/ParameterCore.java:624]: 1. Please add this new parameter to the `kitodo_config.properties` files as this parameter is needed if you use a ldap server. 
2. Add your comment even to this parameter description in the files as this description is important too
3. Maybe rename this parameter to a more meaning full parameter name as the current name "security.secret" could mean a lot but this parameter is the secret for securing the used ldap manager password.

File: Kitodo/src/main/java/org/kitodo/production/forms/LdapServerForm.java [hunk 7]
```diff
@@ -69,9 +73,15 @@ public String newLdapServer() {
...
+            String securitySecret = ConfigCore.getParameterOrDefaultValue(ParameterCore.SECURITY_SECRET);
+
+            if (StringUtils.isNotBlank(securitySecret)) {
+                ldapServer.setManagerPassword( AESUtil.encrypt(ldapServer.getManagerPassword(), securitySecret) );
+            }
+
+            ServiceManager.getLdapServerService().saveToDatabase(ldapServer);
             return usersPage;
-        } catch (DAOException e) {
+        } catch ( Exception e ) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/forms/LdapServerForm.java:84]: Catching every exception here is not a good idea. Fetch only this exception which you would handle here and let all other go on.

File: Kitodo/src/main/java/org/kitodo/production/forms/LdapServerForm.java [hunk 8]
```diff
@@ -121,6 +131,27 @@ public void setLdapServerById(int ldapServerID) {
...
+     */
+    public String getDecryptedManagerPassword() {
+        if (AESUtil.isEncrypted(ldapServer.getManagerPassword())) {
+            String securitySecret = ConfigCore.getParameterOrDefaultValue(ParameterCore.SECURITY_SECRET);
+
+            if (StringUtils.isNotBlank(securitySecret)) {
+                try {
+                    ldapServer.setManagerPassword(AESUtil.decrypt(ldapServer.getManagerPassword(), securitySecret));
+                } catch (Exception e) {
+                    e.printStackTrace();
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/forms/LdapServerForm.java:147]: Using `e.printStackTrace()` is not good. Log this exception if this is needed and you want to handle this exception here.

File: Kitodo/src/main/java/org/kitodo/production/forms/LdapServerForm.java [hunk 9]
```diff
@@ -121,6 +131,27 @@ public void setLdapServerById(int ldapServerID) {
...
+     * @return The decrypted manager password
+     */
+    public String getDecryptedManagerPassword() {
+        if (AESUtil.isEncrypted(ldapServer.getManagerPassword())) {
+            String securitySecret = ConfigCore.getParameterOrDefaultValue(ParameterCore.SECURITY_SECRET);
+
+            if (StringUtils.isNotBlank(securitySecret)) {
+                try {
+                    ldapServer.setManagerPassword(AESUtil.decrypt(ldapServer.getManagerPassword(), securitySecret));
+                } catch (Exception e) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/forms/LdapServerForm.java:146]: Comment like above: catching every exception is not needed - catch only this exception which you will handle here.

File: Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java [hunk 10]
```diff
@@ -534,4 +540,71 @@ public void startNewspaperMigration() {
...
+        }
+
+        List<LdapServer> ldapServers = getLdapServers();
+        ldapServers.parallelStream().forEach(ldapServer -> {
+            String managerPassword = ldapServer.getManagerPassword();
+            if (StringUtils.isNotBlank(managerPassword) && !AESUtil.isEncrypted(managerPassword)) {
+                try {
+                    ldapServer.setManagerPassword(AESUtil.encrypt(managerPassword, securitySecret));
+                    ServiceManager.getLdapServerService().saveToDatabase(ldapServer);
+                } catch (Exception e) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/forms/MigrationForm.java:580]: Same comment like above to general exception catching.

File: Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+     * @return boolean true if value is encrypted using encrypt function
+     */
+    public static boolean isEncrypted(String potentialEncryptedValue) {
+        try {
+            byte[] cipherCombined = Base64.getDecoder().decode(potentialEncryptedValue);
+            byte[] saltPrefix = Arrays.copyOfRange(cipherCombined, 0, SALT_PREFIX.getBytes().length);
+            // check if cipher combined has salt prefix
+            return Arrays.equals(saltPrefix, SALT_PREFIX.getBytes());
+        } catch (IllegalArgumentException e) {
+            // not valid base64 string so
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java:113]: Catching an exception without handling it is bad. At least log this exception in as trace or debug message.

File: Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        Cipher cipher = Cipher.getInstance(ALGORITHM);
+        cipher.init(Cipher.DECRYPT_MODE, getSecretKey(secret, salt), new IvParameterSpec(iv));
+        byte[] value = cipher.doFinal(cipherText);
+        return new String(value);
+    }
+
+    private static SecretKey getSecretKey(String secret, byte[] salt)
+            throws InvalidKeySpecException, NoSuchAlgorithmException {
+        if (StringUtils.isBlank(secret) || Objects.isNull(salt)) {
+            return null;
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/security/AESUtil.java:152]: Avoid returning `null` value as `null` checks must be done everywhere this method is used.

File: Kitodo/src/main/java/org/kitodo/production/services/data/LdapServerService.java [hunk 13]
```diff
@@ -128,7 +130,24 @@ private String buildUserDN(User inUser) {
...
         env.put(Context.SECURITY_PRINCIPAL, ldapServer.getManagerLogin());
-        env.put(Context.SECURITY_CREDENTIALS, ldapServer.getManagerPassword());
+
+        String managerPassword = ldapServer.getManagerPassword();
+        if (AESUtil.isEncrypted(managerPassword)) {
+            String securitySecret = ConfigCore.getParameterOrDefaultValue(ParameterCore.SECURITY_SECRET);
+
+            if (StringUtils.isBlank(securitySecret)) {
+                logger.error("The security.secret parameter was not configured in kitodo_config.properties file.");
+                return null;
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/services/data/LdapServerService.java:140]: Avoid returning `null` value as `null` checks must be done everywhere this method is used.

File: Kitodo/src/main/java/org/kitodo/production/services/data/LdapServerService.java [hunk 14]
```diff
@@ -128,7 +130,24 @@ private String buildUserDN(User inUser) {
...
+            }
+
+            try {
+                managerPassword = AESUtil.decrypt(managerPassword, securitySecret);
+            } catch (Exception e) {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/services/data/LdapServerService.java:145]: Same comment like the other general exception caught: catch only this which you will handle.

File: Kitodo/src/main/java/org/kitodo/production/services/data/LdapServerService.java [hunk 15]
```diff
@@ -128,7 +130,24 @@ private String buildUserDN(User inUser) {
...
+                logger.error(e.getLocalizedMessage(), e);
+                return null;
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/services/data/LdapServerService.java:147]: Avoid returning `null` value as `null` checks must be done everywhere this method is used.

File: Kitodo/src/main/webapp/WEB-INF/templates/includes/system/migration.xhtml [hunk 16]
```diff
@@ -104,6 +111,16 @@
                 update="systemTabView:migrationForm"/>
         </h:panelGroup>
 
+        <h:panelGroup id="ldapManagerPasswordsMigration" rendered="#{MigrationForm.ldapManagerPasswordsMigrationRendered}">
+            <div/>
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/WEB-INF/templates/includes/system/migration.xhtml:115]: Please add some text or explanation here as an empty content is bad.

Screenshot_2021-09-22_11-44-43

**Review by henning-gerhardt — CHANGES_REQUESTED** [review]
Changes are working but there are some code change suggestions.

File: Kitodo/src/main/java/org/kitodo/config/enums/ParameterCore.java (hunk 6 above)
Comment (markusweigelt) [reply Kitodo/src/main/java/org/kitodo/config/enums/ParameterCore.java:624]: For 3. I still need some input: I named the parameter more openly in order to use the secret with other encryption as well. 
Is this a case that does not occur, or should a secret be added for each case.

File: Kitodo/src/main/webapp/WEB-INF/templates/includes/system/migration.xhtml [hunk 17]
```diff
@@ -104,6 +111,16 @@
                 update="systemTabView:migrationForm"/>
         </h:panelGroup>
 
+        <h:panelGroup id="ldapManagerPasswordsMigration" rendered="#{MigrationForm.ldapManagerPasswordsMigrationRendered}">
+            <div style="margin: 10px; width: 50%;"><h:outputText value="#{msgs.migrateLdapManagerPasswordsDesc}:"/></div>
```
Comment (stweil) [inline Kitodo/src/main/webapp/WEB-INF/templates/includes/system/migration.xhtml:115]: Why is there a ":"? The resulting text looks strange:
```
Die Migration verschlüsselt das LDAP Manager Passwort mit dem
"security.secret.ldapManagerPassword" Parameter in der
kitodo_config.properties und überschreibt
das Passwort in der Datenbank. Bestehende Verschlüsselungen werden durch ein erneutes
Ausführen der Migration ignoriert.:
```

Comment (markusweigelt) [reply Kitodo/src/main/webapp/WEB-INF/templates/includes/system/migration.xhtml:115]: Exactly, the ":" appears to be a typo and can be removed.

**Pull request merged**